### PR TITLE
[ONNX] Redesign onnx pass to enable shape type dependent pattern conversion

### PR DESCRIPTION
**Pull request by BowenBao**
With the introduction of ONNX shape inference, shape and type are inferred on the fly as operators get converted from ATen to ONNX when running symbolic function. This resolves the shape/type requirement for the symbolic functions. The pre-onnx passes   however, can not be supported by shape inference, since at that stage the operators in the graph are still ATen operators. 

This PR is to update the design of ONNX pass, to enable a mechanism of capturing subgraphs of ATen operators of certain patterns, and convert them later, when shape/type information of upstream operators are available. 

The new design will require pre-onnx passes that need shape/type to be written in two parts, encapsulation and conversion. 

* The encapsulation part will find the nodes of patterns, like how pre-onnx passes were written previously. But instead of converting the nodes, it will encapsulate them into a sub-block of a new placeholder node. This part is called before onnx pass, so it runs before calling symbolic functions. 

* The conversion part will be called inside the onnx pass. In onnx pass, run_symbolic_func will be called for each node in topological order. When it reaches the placeholder node, the conversion part will be invoked. It will convert the nodes inside the sub-block based on pattern. By that time, it will have shape/type of upstream operators available. After the conversion is complete, the placeholder node will be removed, and nodes inside its sub-block converted. Run_symbolic_func will be called for these nodes, and they will be converted from ATen operator to ONNX operator.

This PR includes several other fixes, listed below.
* replace helper.cpp with onnx_utils.cpp for holding utility functions.
* fix `EraseNumberTypes` on Bool type, the code was outdated that back then Bool type doesn't exist.
* enable onnx shape inference in export with parameter/initializer data.
* other code clean ups.

### CODE REVIEW

File: torch/csrc/jit/passes/onnx.cpp [hunk 1]
```diff
@@ -249,8 +270,8 @@ void BlockToONNX(
 
   // Clone the node and add it to the new graph
   auto cloneNode = [&](Node* node) {
-    auto n_ = ctx.block->appendNode(
-        ctx.block->owningGraph()->createClone(node, envFn));
+    auto n_ = new_block->appendNode(
+        new_block->owningGraph()->createClone(node, envFn));
     for (size_t i = 0; i < node->outputs().size(); i++) {
       // n_->outputs()[i]->setType(node->outputs()[i]->type());
       env[node->outputs()[i]] = n_->outputs()[i];
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx.cpp:277]: nit: some inconsistencies, using both outputs().at(i) and outputs()[i]

File: torch/csrc/jit/passes/onnx.cpp [hunk 2]
```diff
@@ -176,10 +176,41 @@ void BlockToONNX(
...
+  }
+
+  // Finally, visit all nodes in the graph
+  for (auto node : old_block->nodes()) {
+    NodeToONNX(node, ctx.block, operator_export_type, env);
+  }
+  for (auto output : old_block->outputs()) {
+    ctx.block->registerOutput(env.at(output));
+  }
+  EliminateDeadCode(
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx.cpp:197]: This was original code moved up. Adding the comment to make it clearer.

File: torch/csrc/jit/passes/onnx.cpp (hunk 1 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx.cpp:277]: i discovered that we could use `output(i)` as it turns out to be a checked access, calling `outputs_.at(i)` underneath. i'm updating this part with this change.

File: torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp [hunk 3]
```diff
@@ -332,29 +310,60 @@ void ONNXFixupUninitializedOutput(Node* node) {
...
+
+void FixupONNXLoopNodeInputs(Node* node) {
+  if (node->kind() != ::c10::onnx::Loop) {
+    return;
+  }
+
+  auto* graph = node->owningGraph();
+
+  // add cast to condition input outside the loop.
+  Value* cond_val = node->inputs()[1];
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:348]: same above, changing to `input(1)` for checked access.

File: torch/csrc/jit/passes/onnx/pattern_conversion/pattern_conversion.cpp [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+    const std::vector<Node*>& slice_and_select_nodes,
+    Value* orig_data,
+    const std::unordered_map<Value*, Value*>& env) {
+  std::unordered_map<int64_t, ConvertedIndex> dim_index_map;
+
+  // Loop over fetched slice and select nodes and convert them to index tensors.
+  // keep track of which dimension the current slice/select node is applying to.
+  int64_t cur_dim = 0;
+  int64_t dim_offset = 0;
+  const auto orig_tensor_indices = index_put_node->input(1)->node()->inputs();
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/pattern_conversion/pattern_conversion.cpp:100]: ditto

**Comment by BowenBao** [comment]
closing to move to merge with dev branch

**Pull request closed without merging**